---
title: "Christopher's Awesome Problem Set 3"
author: "Chris Erickson"
date: "Feb 23, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#>")
```

# Submission

1. Fork this repository to your own account
1. Make changes to the `README.Rmd` file (including the author field above).
1. Knit the file to HTML format and add it to your git repository (`git add`)
1. Submit a pull request with your Rmd and knitted HTML files.

# Overview 

You will examine a [time-course of mRNA expression][1] of barley powdery mildew fungus in immunocompromised plants (*Arabidopsis*). The fungus Blumeria graminis f. sp. hordei (called *Bgh* from now on)

Counts were generated using the [`maSigPro` software][3] and deposited in [NCBI GEO][2].

[1]: http://www.ncbi.nlm.nih.gov/pubmed/23175786
[2]: http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE43163  
[3]: http://bioinformatics.oxfordjournals.org/content/30/18/2598.full

Some important identifiers for the data set:

| Name | Description |
| ---- | ----------- |
| `A6` | *Bgh* isolate expressing other AVRA effectors |
| `K1` | *Bgh* isolate expressing the cognate AVRA1 effector for MLA1|
| `pps` | *Arabidopsis* plants |
| `B12` | *Arabidopsis* plants expressing MLA1-HA|

We will download tables from GEO containing counts of mRNA abundance from both the fungal sample.

# Raw data

First we need load the data.

```{r load_raw_data, message = FALSE}
library(tidyverse)
library(dplyr)
library(stringr)
library(tidyr)

# bgh data
bgh_url <- 'http://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE43163&format=file&file=GSE43163_CompleteCountTable_Bgh.txt.gz'

# 2 info and blank lines at top, skip them
raw_data <- read_tsv(bgh_url, skip = 2)

# the header for the first column is "NA", set it manually
names(raw_data)[1] <- 'gene.name'

raw_data <- as_data_frame(raw_data) #Converts into tibble

```

# Problems


## Problem 1

Now that the raw data are loaded, your first task is to tidy the data with `tidyr`. The `cleaned_data` should look something like this:

```
# A tibble: 310,896 Ã— 6
             gene.name plant fungus time.point   rep value
*                <chr> <chr>  <chr>      <chr> <chr> <int>
1             bgh04079   B12     A6       6hpi     1    13
2             bgh01634   B12     A6       6hpi     1    31
3  bghG000012000001001   B12     A6       6hpi     1   121
4  bghG000012000002001   B12     A6       6hpi     1     3
5             bgh00757   B12     A6       6hpi     1   253
6             bgh01273   B12     A6       6hpi     1    45
7             bgh01274   B12     A6       6hpi     1    32
8             bgh01277   B12     A6       6hpi     1     9
9             bgh06140   B12     A6       6hpi     1    47
10            bgh05774   B12     A6       6hpi     1    42
# ... with 310,886 more rows
```

The key steps are to `gather()` the data into key-value pairs and then `separate()` the information from the key column into new columns.

```{r problem_1}

#Note raw_data has rows of gene.names, and columns of each observation. It has plantname_fungusname_timepoint_replicate, and this needs to be broken up into their own variables.

# Use gather() to take multiple columns and collape them into key-value pairs. Gather() moves wide data into long data. Will output a key column and a value column based on the genes.name column. How does it know???????????????

# Separate does the opposite, turns long data into wide. In this case, the concatenated data separated by '_' is separated into 4 columns.

raw_data %>% gather(key, value, -gene.name) %>% separate(key, into = c('plant', 'fungus', 'time.point', 'rep'), sep = '_') -> cleaned_data

```

## Problem 2

You need to translate what you learned on the command line into the R world.

Which plant has the highest expression of any gene in the 6hpi time point?

```{bash, eval=FALSE}
raw_data | awk '$4 == "6hpi"' | sort -k6nr | head -n 1 | cut -f2
```

```{r}

#First, filter() will select the rows from the data.table with the 6hpi timepoint. Then, data will be arranged so that the gene with the highest expression will be on top, using arrange() on the 'value' variable.

filter(cleaned_data, time.point == "6hpi") %>% arrange(desc(value)) -> highest

#The plant with the highest gene expression will then be selected.

head(highest, n = 1L) %>% select(plant)

```

Which plant / fungus pair has the highest expression in the 18hpi time point?

```{bash, eval=FALSE}
raw_data | awk '$4 == "18hpi"' | sort -k6nr | head -n 1 | cut -f2,3
```

```{r}
#To try something different, first filter the data to select the 18hpi timepoint as above. Then arrange the data so that the gene with the highest expression (value) will be first using arrange().

filter(cleaned_data, time.point =="18hpi") %>% arrange(desc(value)) %>% head(n = 1L) %>% select(plant, fungus)

```


## Problem 3

Identify the top 3 most consistently differentially expressed genes between the earliest and latest time points for each combination of `plant` and `fungus` strains.

- "Differential expression"" is the difference between `value` (i.e., gene expression level) between time points.

- "Consistency" is the smallest variance in `value` between replicates.

### Strategy

1. Create a new table from the cleaned data by moving each `hpi` value to a new column name, with counts for each in the column (hint: use a `tidyr` verb). It is helpful to reformat the `hpi` values by converting from e.g. `6hpi` to `hpi.6`. You can use `mutate` to do this, i.e.:

```{r reformat_time}
#First, add two new columns with mutate(): one named time.value which is the time.point column but without the 'hpi'. Use str_replace to replace 'hpi' with nothing ''. Second one is named time.hpi, and concatenates 'hpi.' to the time.value column giving 'hpi.6.' This separation makes data manipulation easier in the long run.

library(stringr)

cleaned_data %>%
  mutate(time.value = str_replace(time.point, 'hpi', ''),
         time.hpi = str_c('hpi.', time.value)) -> cleaned_data_1

#Then we want to create new columns (variables) for each of the timepoints. In time.point, there are 4 timepoints 6,12,18,24hpi but they are all in this single time.point column. We want each time point to be its own variable, and can use 'spread' to do this. Spread does the opposite of gather, it makes long data wide

#Note: the time.point and time.value columns must be removed before spreading in order to 'collapse' the values for each time point onto a single row. If not, each timepoint will have a value on its own row, so then can't subtract 24h-6h; values will be 'NA'. Show every column except time.point and time.value by using select(-columnnametojettison). 

cleaned_data_1 %>% select(-time.point, -time.value) %>% spread(time.hpi, value) -> cleaned_data_2

```

2. Create a new column containing the expression difference between the relevant time points.

3. Calculate summary statistics (mean and variance) of the expression differences by grouping (hint) the `gene.name`, `plant`, and `virus` columns.

4. Sort by these statistics and use the `dplyr` verb `slice` to pull the ones you want (i.e., the top 3). Note you will have to remove `gene.name` from the grouping so that sorting works.

```{r problem_3}

# Now differences between the values from first and last timepoints, 24hr-6hr, can be analyzed. Add a new column giving this difference, expr_data, with mutate().

cleaned_data_2 %>% mutate(expr.diff = hpi.24 - hpi.6) -> expr_data

# Calculate the summary statistics (mean and variance) using summarise(), after grouping with group_by(). Grouping here makes groups of plants, fungi, and gene names. Makes 3 groups????? Its commonly useful to report summary statistics on groups of data, rather than on the whole data set. Grouping and then summarizing will do just that. 

expr_data %>% group_by(plant, fungus, gene.name) %>% summarise(meanval = mean(expr.diff), expr.var = var(hpi.24), expr.diff.var = var(expr.diff)) -> expr_data_sumd

#Now that we have the expression difference (expr.diff) and the vairance (expr.diff.var) we can use select(for columns) and filter(for rows) to identify the top 3 gene names with the highest expression difference (highest mean(expr.diff)) and lowest variance (lowest expr.diff.var). Do this using arrange() by meanval and then the variance.

arrange(expr_data_sumd, desc(meanval), expr.diff.var) -> expr_data_sumd

head(expr_data_sumd, n=3) %>% select(gene.name) -> top3

top3

```

## Problem 4

Now examine the above final data frame above and write a few sentences putting inline code in least 3 places. For example, There are `r nrow(mtcars)` rows of data in `mtcars`.

This analysis in problem 3 finds the highest difference in gene expression across the time points from 6 to 24hrs with the lowest variance. 

START: `r nrow(raw_data)` genes were analyzed in `r length(unique(cleaned_data$fungus))*length(unique(cleaned_data$plant))` plant-fungi combos. Of these genes, the top 3 most consistently differentially expressed were `r ungroup(top3) %>% select(gene.name) %>% slice(1)`, `r ungroup(top3) %>% select(gene.name) %>% slice(2)`, and `r ungroup(top3) %>% select(gene.name) %>% slice(3)`.

## Problem 5

Plot the expression (`value`) by time (`hpi`) for the above genes. Format the plot as follows:

1. Plot each point.
1. Connect the points with a smooth line 
1. Create a facet for each `plant` / `bgh` pair 
1. Assign a different color to each gene 
1. Assign a different shape to each replicate 
1. Add a descriptive title to the plot 
1. Use the "black & white" theme

Extra credit: add error bars to the plot (use `geom_errorbar`).

```{r problem_5}

filter(cleaned_data, gene.name == "bgh02634" | gene.name == "bgh06450" | gene.name == "bghG000349000001001") -> plotme

library(ggplot2)

ggplot(plotme, aes(time.point, value)) + geom_point()

```
